File: omnitiles/src/protocol/messages.rs
// SPDX-License-Identifier: MIT
// © 2025–2026 Christopher Liu

//! Command message protocol used to communicate with OmniTiles.

/// Sync byte for the protocol.
pub const START_BYTE: u8 = 0xA5;

// Message IDs
pub const MSG_M1_EXTEND: u8 = 0x30;
pub const MSG_M1_RETRACT: u8 = 0x31;
pub const MSG_M1_BRAKE: u8 = 0x32;
pub const MSG_M1_SET_POSITION: u8 = 0x33;

pub const MSG_M2_EXTEND: u8 = 0x40;
pub const MSG_M2_RETRACT: u8 = 0x41;
pub const MSG_M2_BRAKE: u8 = 0x42;
pub const MSG_M2_SET_POSITION: u8 = 0x43;

pub const MSG_PING: u8 = 0x50;

pub const MSG_TELEMETRY: u8 = 0x60;

/// Direct motor commands.
#[derive(Debug, Clone, Copy, PartialEq)]
pub enum Command {
    Ping,
    M1Extend(u8),
    M1Retract(u8),
    M1Brake,
    M1SetPosition(u8),
    M2Extend(u8),
    M2Retract(u8),
    M2Brake,
    M2SetPosition(u8),
}

File: sdk/src/omnitiles/protocol/messages.py
"""Binary protocol message IDs.

Mirrors ``omnitiles/src/protocol/messages.rs`` — keep these in sync with the
firmware. Packet format:

    [START_BYTE] [msg_id] [payload...] [checksum]

``checksum`` is the 8-bit sum of ``msg_id`` and all payload bytes.
"""

from enum import IntEnum

START_BYTE: int = 0xA5


class MessageId(IntEnum):
    """Protocol message identifiers."""

    M1_EXTEND = 0x30
    M1_RETRACT = 0x31
    M1_BRAKE = 0x32
    M1_SET_POSITION = 0x33

    M2_EXTEND = 0x40
    M2_RETRACT = 0x41
    M2_BRAKE = 0x42
    M2_SET_POSITION = 0x43

    PING = 0x50

    TELEMETRY = 0x60

    BASE_VELOCITY = 0x70
    BASE_BRAKE = 0x71
